server/src/main/java/io/druid/segment/realtime/RealtimeManager.java

```diff
@@ -319,6 +299,128 @@ public void run()
+      }
+      catch (Exception e) {
+      	log.error(e, "Failed to start firehoseV2");
+      	return;
+      }
+      long nextFlush = new DateTime().plus(intermediatePersistPeriod).getMillis();
+
+      boolean haveRow = true;
+      while (haveRow) {
+        InputRow inputRow = null;
+        int numRows = 0;
+        try {
+          inputRow = firehose.currRow();
+          if (inputRow != null) {
+            try {
+              numRows = plumber.add(inputRow);
+            }
+            catch (IndexSizeExceededException e) {
+              log.info("Index limit exceeded: %s", e.getMessage());
+              nextFlush = doIncrementalPersist(firehose.makeCommitter(), intermediatePersistPeriod);
+              continue;
+            }
+
+            if (numRows < 0) {
+              metrics.incrementThrownAway();
+              log.debug("Throwing away event[%s]", inputRow);
+            } else {
+              metrics.incrementProcessed();
+            }
+          }
+        }
+        catch (Exception e) {
+          log.makeAlert(e, "Some exception got thrown while processing rows.  Ignoring and continuing.")
+             .addData("inputRow", inputRow);
+        }
+
+        try {
+          haveRow = firehose.advance();
+        }
+        catch (Exception e) {
+          log.debug(e, "thrown away line due to exception, considering unparseable");
+          metrics.incrementUnparseable();
+          continue;
+        }
+
+        try {
+          if (numRows > config.getMaxRowsInMemory() || System.currentTimeMillis() > nextFlush) {
+            nextFlush = doIncrementalPersist(firehose.makeCommitter(), intermediatePersistPeriod);
+          }
+        } catch (Exception e) {
+          log.makeAlert(e, "An exception happened while queue to persist!?  We hope it is transient. Ignore and continue.");
+        }
+      }
+    }
+
+    private long doIncrementalPersist(Committer committer, Period intermediatePersistPeriod)
+    {
+      plumber.persist(committer);
+      return new DateTime().plus(intermediatePersistPeriod).getMillis();
+    }
+
+
+    private void runFirehose(Firehose firehose)
+    {
+
+      final Period intermediatePersistPeriod = config.getIntermediatePersistPeriod();
+
+      long nextFlush = new DateTime().plus(intermediatePersistPeriod).getMillis();
+
+      while (firehose.hasMore()) {
+        InputRow inputRow = null;
+        try {
+          try {
+            inputRow = firehose.nextRow();
+          }
+          catch (Exception e) {
+            log.debug(e, "thrown away line due to exception, considering unparseable");
+            metrics.incrementUnparseable();
+            continue;
+          }
+
+          boolean lateEvent = false;
+          boolean indexLimitExceeded = false;
+          try {
+            lateEvent = plumber.add(inputRow) == -1;
+          }
+          catch (IndexSizeExceededException e) {
+            log.info("Index limit exceeded: %s", e.getMessage());
```

Comment:
Is there a reason why they can't actually use the same logic for both kinds of Firehoses? Like, do we really need to throw away rows that triggered IndexSizeExceededException in the firehose v1 code?